src/java.base/share/classes/sun/net/www/MimeTable.java

```diff
@@ -151,28 +154,56 @@ public synchronized MimeEntry find(String type) {
     }
 
     /**
-     * Locate a MimeEntry by the file extension that has been associated
-     * with it. Parses general file names, and URLs.
+     * Returns the file extension or the empty string if none found.
      */
-    public MimeEntry findByFileName(String fname) {
-        String ext = "";
-
-        int i = fname.lastIndexOf('#');
-
-        if (i > 0) {
-            fname = fname.substring(0, i - 1);
-        }
-
-        i = fname.lastIndexOf('.');
+    private static String getFileExtension(String fname) {
+        int i = fname.lastIndexOf('.');
         // REMIND: OS specific delimiters appear here
         i = Math.max(i, fname.lastIndexOf('/'));
         i = Math.max(i, fname.lastIndexOf('?'));
 
+        String ext = "";
         if (i != -1 && fname.charAt(i) == '.') {
             ext = fname.substring(i).toLowerCase();
         }
 
-        return findByExt(ext);
+        return ext;
+    }
+
+    /**
+     * Locate a MimeEntry by its associated file extension.
+     * Parses general file names, and URLs.
+     *
+     * @param fname the file name
+     *
+     * @return the MIME entry associated with the file name or {@code null}
+     */
+    public MimeEntry findByFileName(String fname) {
+        MimeEntry entry = null;
+
+        // If an optional fragment introduced by a hash mark is
+        // present, then strip it and use the prefix
+        int hashIndex = fname.lastIndexOf(HASH_MARK);
+        if (hashIndex > 0) {
+            String ext = getFileExtension(fname.substring(0, hashIndex - 1));
```

Comment:
Hello Brian, I think there's a bug here. Not introduced by this PR but even in the current master. I think that `fname.substring(0, hashIndex -1)` call should actually be `fname.substring(0, hashIndex)`. For example, in its current form, if `fname` is `a.png#foo` then `fname.substring(...)` here will return `a.pn` instead of `a.png`. It looks like we don't have a test for that case.